merkledag/ipld.md

```diff
@@ -0,0 +1,464 @@
+# IPLD -- the "thin-waist" merkle dag format.
+
+There are a variety of systems that use merkle-tree and hash-chain inspired datastructures (e.g. git, bittorrent, ipfs, tahoe-lafs, sfsro). IPLD defines:
+
+- **_merkle-links_**: the core unit of a merkle-graph
+- **_merkle-dag_**: any graphs whose edges are _merkle-links_.
+- **_merkle-paths_**: unix-style paths for traversing _merkle-dags_ with _named merkle-links**
+- **IPLD Data Model**: a flexible, JSON-based data model for representing merkle-dags.
+- **IPLD Serialized Formats**: a set of formats in which IPLD objects can be represented, for example JSON, CBOR, CSON, YAML, Protobuf, XML, RDF, etc.
+- **IPLD Canonical Format**: a deterministic description on a serialized format that ensures the same _logical_ object is always serialized to _the exact same sequence of bits_. This is critical for merkle-linking, and all cryptographic applications.
+
+In short: JSON documents with named merkle-links that can be traversed.
+
+## Intro
+
+### What is a _merkle-link_?
+
+A _merkle-link_ is a link between two objects which is content-addressed with the _cryptographic hash_ of the target object, and embedded in the source object. Content addressing with merkle-links allows:
+
+- **Cryptographic Integrity Checking**: resolving a link's value can be tested by hashing. In turn, this allows wide, secure, trustless exchanges of data (e.g. git or bittorrent), as others cannot give you any data that does not hash to the link's value.
+- **Immutable Datastructures**: data structures with merkle links cannot mutate, which is a nice property for distributed systems. This is useful for versioning, for representing distributed mutable state (eg CRDTs), and for long term archival.
+
+### What is a _merkle-graph_ or a _merkle-dag_?
+
+Objects with merkle-links form a Graph (merkle-graph), which necessarily is both Directed, and which can be counted on to be Acyclic, iff the properties of the cryptographic hash function hold. I.e. a _merkle-dag_. Hence all graphs which use _merkle-linking_ (_merkle-graph_) are necessarily also Directed Acyclic Graphs (DAGs, hence _merkle-dag_).
+
+### What is a _merkle-path_?
+
+A _merkle-path_ is a unix-style path (e.g. `/a/b/c/d`) which initially dereferences through a _merkle-link_ and then follows _named merkle-links_ in the intermediate objects. Following a name means looking into the object, finding the _name_ and resolving the associated _merkle-link_.
+
+For example, suppose we have this _merkle-path_:
+
+```
+/ipfs/QmUmg7BZC1YP1ca66rRtWKxpXp77WgVHrnv263JtDuvs2k/a/b/c/d
+```
+Where:
+- `ipfs` is a protocol namespace (to allow the computer to discern what to do)
+- `QmUmg7BZC1YP1ca66rRtWKxpXp77WgVHrnv263JtDuvs2k` is a cryptographic hash.
+- `a/b/c/d` is a path _traversal_, as in unix.
+- this link traverses five objects.
```

Comment:
I would not expect this to be an issue because what you describe is the compact form of the following object:

    > ipld cat --fmt yml $h1
    ---
    foo: 
      mlink: $h2
      bar:
        mlink: $h3

And in this object, only the "foo" link will be considered valid, not "foo/bar" because :

- keys should not be allowed to have `/` character (so we will never have "foo/bar" verbatim in a key) for the same reasons `/` is not allowed in a unix filename.

- in the previous object, "foo/bar" will not be considered as a link as per the  "Duplicate property keys" section :

    >   Note that having two properties with the same name IS NOT ALLOWED, but actually impossible to prevent (someone will do it and feed it to parsers), so to be safe, we define the value of the path traversal to be the first entry in the serialized representation. For example, suppose we have the object: